pkg/kubelet/cm/cpumanager/state/state_checkpoint.go

```diff
@@ -0,0 +1,186 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package state
+
+import (
+	"fmt"
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/kubelet/checkpointmanager"
+	"k8s.io/kubernetes/pkg/kubelet/checkpointmanager/errors"
+	"k8s.io/kubernetes/pkg/kubelet/cm/cpuset"
+	"path"
+	"sync"
+)
+
+// cpuManagerCheckpointName is the name of checkpoint file
+const cpuManagerCheckpointName = "cpu_manager_state"
+
+var _ State = &stateCheckpoint{}
+
+type stateCheckpoint struct {
+	sync.RWMutex
+	policyName        string
+	cache             State
+	checkpointManager checkpointmanager.CheckpointManager
+}
+
+// NewCheckpointState creates new State for keeping track of cpu/pod assignment with checkpoint backend
+func NewCheckpointState(stateDir string, policyName string) (State, error) {
+	checkpointManager, err := checkpointmanager.NewCheckpointManager(stateDir)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize checkpoint manager: %v", err)
+	}
+	stateCheckpoint := &stateCheckpoint{
+		cache:             NewMemoryState(),
+		policyName:        policyName,
+		checkpointManager: checkpointManager,
+	}
+
+	if err := stateCheckpoint.tryRestoreState(); err != nil {
+		return nil, fmt.Errorf("unable to restore state from checkpoint: %v\n"+
+			"Please drain this node and delete the CPU manager checkpoint file \"%s\" before restarting Kubelet.",
+			err, path.Join(stateDir, cpuManagerCheckpointName))
+	}
+
+	return stateCheckpoint, nil
+}
+
+// tryRestoreState tries to read checkpoint and creates it if it doesn't exist
+func (sc *stateCheckpoint) tryRestoreState() error {
+	sc.Lock()
+	defer sc.Unlock()
+	var err error
+
+	// used when all parsing is ok
+	tmpAssignments := make(ContainerCPUAssignments)
+	tmpDefaultCPUSet := cpuset.NewCPUSet()
+	tmpContainerCPUSet := cpuset.NewCPUSet()
+
+	checkpoint := NewCPUManagerCheckpoint()
+	if err = sc.checkpointManager.GetCheckpoint(cpuManagerCheckpointName, checkpoint); err != nil {
+		if err == errors.ErrCheckpointNotFound {
+			sc.storeState()
+			return nil
+		}
+		return err
+	}
+
+	readState := *(checkpoint.(*cpuManagerCheckpoint))
```

Comment:
* make sure that outside deference is still needed
* handle type conversion errors.